source/reduce/change_operand_to_undef_reduction_opportunity.cpp

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/reduce/change_operand_to_undef_reduction_opportunity.h"
+
+#include "source/opt/ir_context.h"
+#include "source/reduce/reduction_util.h"
+
+namespace spvtools {
+namespace reduce {
+
+bool ChangeOperandToUndefReductionOpportunity::PreconditionHolds() {
+  // Check that the instruction still has the original operand.
+  return inst_->NumOperands() > operand_index_ &&
+         inst_->GetOperand(operand_index_).words[0] == original_id_;
+}
+
+void ChangeOperandToUndefReductionOpportunity::Apply() {
+  auto operand = inst_->GetOperand(operand_index_);
+  auto operand_id = operand.words[0];
```

Comment:
Yes. However, I feel that the opportunity should be thought of as consisting of the Instruction `inst_`, plus the `operand_index_`. The other fields are for verifying that the opportunity can still be applied, and should, in general, not be necessary for applying the opportunity and so should perhaps not be accessed here (and could be changed in the future, although it is fairly trivial in this case). At least, this is how I think of it.

Regardless, perhaps we should be doing `assert(PreconditionHolds())` in all implementations of `Apply()`, since that is supposed to be a precondition of this function. On the other hand, the only way to call the protected `Apply()` function is via `ReductionOpportunity::TryToApply()`, which checks the precondition.

Reply:
Okay that makes sense.  You can leave it as it.